### PR TITLE
Fix cuda and cudnn libraries search process on Windows

### LINKED ISSUE #20202: cuDNN arch mismatch - software or hardware error?
Hello, I had a lot of fun with training neural networks on my GPU but then I tried LSTM.

```Python
    x = self.lstm(x)
  File "D:\Anaconda3\lib\site-packages\torch\nn\modules\module.py", line 489, in __call__
    result = self.forward(*input, **kwargs)
  File "D:\Anaconda3\lib\site-packages\torch\nn\modules\rnn.py", line 179, in forward
    self.dropout, self.training, self.bidirectional, self.batch_first)
RuntimeError: cuDNN error: CUDNN_STATUS_ARCH_MISMATCH
```

My GPU is `GeForce GTX 950` and it's `NVIDIA Compute Capability` is `5.2`.

```
>>> import torch
>>> torch.backends.cudnn.version()
7501
```
I would really appreciate quick guess, is the hardware too old or could that be an environment issue?

### PR DESCRIPTION
**Pull request by peterjc123**
Fixes #20202

### CODE REVIEW

File: torch/backends/cudnn/__init__.py
Comment (ezyang) [inline torch/backends/cudnn/__init__.py]: I was going to complain that this isn't portable but this is Windows only :>

Comment (ezyang) [reply torch/backends/cudnn/__init__.py]: I see this code is substantially duplicated for cuda as well. Can we make a helper function to do this instead?

**Review by ezyang — APPROVED** [review]
This seems OK, but I'd prefer for the user to confirm this fixes their issue before landing.

**Comment by ezyang** [comment]
Also, in the implementation where we check lib first, can you please cite the issue in question this solves. That will help people in the future trying to understand what the issue is.

File: torch/backends/cudnn/__init__.py
Comment (ezyang) [reply torch/backends/cudnn/__init__.py]: It would logically  make sense to me if cudnn imported cuda. Does that result in a circular import?

Comment (soumith) [reply torch/backends/cudnn/__init__.py]: i'm not looking forward to de-copy/paste unless it's necessary because this code is very specific and possibly will change with cudnn / cuda versions.

Comment (peterjc123) [reply torch/backends/cudnn/__init__.py]: One is in `torch/cuda/__init__.py` and the other one is in `torch/backends/cudnn/__init__.py`. And currently, there is no dependency between these two scripts. A safe and easy option is to put the function inside one file under the root directory of torch.

Comment (soumith) [reply torch/backends/cudnn/__init__.py]: let's just leave them as-is for now

Comment (peterjc123) [reply torch/backends/cudnn/__init__.py]: @soumith Sure, I'm okay with this.

**Comment by ezyang** [comment]
OK, I'm happy to defer to @soumith here

Excerpts from Soumith Chintala's message of 2019-05-07 08:10:14 -0700:
> soumith commented on this pull request.
> 
> > @@ -27,7 +27,19 @@
>  
>  
>  def find_cudnn_windows_lib():
> -    proc = Popen(['where', 'cudnn64*.dll'], stdout=PIPE, stderr=PIPE, stdin=PIPE)
> +    # Override the default search process
> +    # The libary selection will be done in these directories one by one
> +    # 1. [Package Root]\Lib 
> +    #    That's where our libraries are in, which should be loaded first.
> +    # 2. Default directories
> +    #    That is stored in the environment variable `PATH`.
> +    test_env = os.environ.copy()
> +    old_path = test_env['PATH']
> +    th_dll_path = os.path.join(os.path.dirname(
> +        os.path.dirname(os.path.dirname(__file__))), 'lib')
> +    test_env['PATH'] = ';'.join([th_dll_path, old_path])
> 
> let's just leave them as-is for now
>

**Comment by crackcomm** [comment]
I can confirm this patch totally eliminated issue #20202.

**Comment by peterjc123** [comment]
@ezyang @soumith Done!

**Pull request closed without merging**
